### ISSUE TITLE
too big subrequest response

### ISSUE THREAD

**Issue by bit-111** [kaltura/nginx-vod-module#1060]
hello,

I use the vod module and I got an error while trying to serve vod to the client.
> [error] 56084#56084: *3 too big subrequest response while sending to client, client: 127.0.0.1 .... subrequest: "/generate_json/vod/.....", upstream: "fastcgi://127.0.0.1:10063", host: "127.0.0.1:10062"

port 10063 is the generate json and port 10062 is for the vod request.
The subrequest is regarding generating json file to make a playlist for the vod and this error only comes out if I tried to serve vod from more than certain number of videos (e.g. more than 20 videos, it will give this error. If i use only 5 videos, it will successfully serve it). After searching the forums, I came to conclussion that it has something to do with `subrequest_output_buffer_size`. I think the subrequest response (json file) is too big for the buffer. Then I tried to increase the buffer size by using this syntax in nginx.conf file.

# http
        server {
        include mime.types;
                listen 80;
                server_name _;
                root /var/www/html;

                index index.php index.html;

                location / {
                        try_files $uri $uri/ =404;
                }

                location ~ \.php$ {
            #try_files $uri =404;
        #    root           html;
                include fastcgi-php.conf;

            #fastcgi_pass unix:/var/run/php/php7.2-fpm.sock;
            fastcgi_pass   127.0.0.1:9000;

            #fastcgi_index  index.php;

            #fastcgi_param  SCRIPT_FILENAME  /scripts$fastcgi_script_name;
            include        fastcgi_params;
        }

        location ~ /\.ht {
            deny all;
        }
        }

        # Optimization settings
    ssi on;
    subrequest_output_buffer_size 128k;
    sendfile on;
    tcp_nopush on;
    tcp_nodelay on;
    aio on;
    directio 512;
    keepalive_timeout 60;`

But until now, I have no luck with the error. I am wondering if there is something wrong with my command, or is it not possible to set the `subrequest_output_buffer_size` bigger?

**Comment by erankor** [comment]
Try to increase `vod_max_mapping_response_size` - when this module creates subrequests, it allocates the response buffer, not nginx core. 
However, as far as I understand (you didn't paste the whole conf...) you are pointing vod_upstream_location to a location that uses fastcgi_pass. While it may work, you should note I only tested it with proxy_pass...

**Comment by bit-111** [comment]
Thank you very much for your speedy answer. It works well with my code. But now, how can I calculate the size of the buffer I need, for example 100 or 200 videos? Do you have any tips?

**Comment by erankor** [comment]
It depends on the specific JSON format you use, and probably mostly on the length of the MP4 paths. 
I think easiest is to try several options (e.g. 50, 100 etc.), by curl-ing the JSON and looking at the content-length, from that, you can extrapolate the size you need (you would also need to add some small amount to account for the HTTP headers).

**Comment by bit-111** [comment]
okay, so by looking at my JSON file for specific number of videos I can predict the buffer size. Do you know if it is scaling linearly (e.g. 5 videos = 1k, then 100 videos = 20k)? And what is the safe upper limit for this buffer size? can i just increase it to several mega?

**Comment by erankor** [comment]
Assuming the path lengths are fairly uniform, yes, it will increase linearly.
Note that the size you use will be allocated for each request - the buffer is allocated in advance before the content-length is known. So probably not a good idea to use a number significantly higher than what you actually need.

**Comment by bit-111** [comment]
Thank for the warning, but i actually need that many amount of videos to load, and the number is almost always the same.
So, i have tried experimenting on buffer size according to numbers of video and I found that there is actually a limit on how big the buffer size is. Even when I set the buffer size bigger, it still will not load videos over 130 videos. Do you know any limiting factor regarding that?

**Comment by erankor** [comment]
Yes, there is a hard limit in the code of 128 clips. There are certain operations that the module has to do per-clip, that's the reason for this limit.

**Comment by bit-111** [comment]
Okay, that is good to know. Thank you very much for your help. I really appreciate it.

**Comment by bit-111** [comment]
By the way, is there any way to increase that in my nginx? or do you have any tips to work around it?

**Comment by erankor** [comment]
If you compile the module yourself, you can easily patch it - change  `#define MAX_CLIPS (128)`

**Comment by bit-111** [comment]
is there any performance consequence if I raise the max clips too much? I am planning to raise to to 750.

**Comment by erankor** [comment]
No, unlike the buffer size, in this case, what matters is the number of clips you actually use
